### PR TITLE
[Colossus] Sync rework - 

### PR DESCRIPTION
**Pull request by zeeshanakram3**
colses #5008

This PR:
- removes `api` parameter from the `performSync` function & fetch the object hash from QN instead of fetching from the runtime.
- directly download data objects from remote nodes instead of pre-checking their existence first.
- The logging message in the `superagent` error handler - when the object does not exist in the remote node - won't be logged, since technically missing asset won't be categorized as an error, and logging this as an error may confuse the Operator/Lead

### CODE REVIEW

File: storage-node/src/services/sync/tasks.ts [hunk 1]
```diff
@@ -77,39 +72,62 @@ export class DeleteLocalFileTask implements SyncTask {
...
   description(): string {
-    return `Sync - downloading file: ${this.url} to ${this.uploadsDirectory} ....`
+    return `Sync - Trying for download of: ${this.dataObjectId} ....`
   }
 
   async execute(): Promise<void> {
+    // Create an array of operator URL indices to maintain a random URL choice
+    // cannot use the original array because we shouldn't modify the original data.
+    // And cloning it seems like a heavy operation.
+    const operatorUrlIndices: number[] = [...Array(this.operatorUrls.length).keys()]
```
Comment (mnaamani) [inline storage-node/src/services/sync/tasks.ts]: I can suggest a slightly cleaner approach. Start with an array already containing random order of the indices using a combination of _.range() and _.shuffle(). You could then just shift or pop the array at the start of the loop. Or the while loop then can be a simply a `for..each` iteration over the array, without needing to modify it.

`const operatorUrlIndices: number[] = _.shuffle(_.range(this.operatorUrls.length))`

**Review by mnaamani — CHANGES_REQUESTED** [review]
Looks great. Small code style change suggested.

File: storage-node/src/services/sync/tasks.ts [hunk 2]
```diff
@@ -77,39 +72,53 @@ export class DeleteLocalFileTask implements SyncTask {
...
 
   async execute(): Promise<void> {
+    const operatorUrlIndices: number[] = _.shuffle(_.range(this.operatorUrls.length))
+
+    for (const randomUrlIndex of operatorUrlIndices) {
+      const chosenBaseUrl = this.operatorUrls[randomUrlIndex]
+      logger.debug(`Sync - random storage node URL was chosen ${chosenBaseUrl}`)
+
+      // Remove random url from the original list.
+      _.remove(operatorUrlIndices, (index) => index === randomUrlIndex)
```
Comment (mnaamani) [inline storage-node/src/services/sync/tasks.ts]: removing the element is no longer necessary, even potentially problematic.

Comment (zeeshanakram3) [reply storage-node/src/services/sync/tasks.ts]: Ah right. Missed that.

Comment (zeeshanakram3) [reply storage-node/src/services/sync/tasks.ts]: Addressed in [3c02240](https://github.com/Joystream/joystream/pull/5014/commits/3c02240ef9fcf9a64945e94c880f33a16849de16)

**Review by mnaamani — APPROVED** [review]

**Pull request merged**